### ISSUE TITLE
Implementing Undo/Redo collides with devtools

### ISSUE THREAD

**Issue by FredyC** [gaearon/redux-devtools#113]
This is another case of actions that are being replayed with devtools which causes inconsistency in history used for undo/redo operation. Here is what I've invented so far...

    export function createHistory(limit = 100) {

        const recordedActions = [
            EditorActionType.SET_ROW_TEXT
        ];

        let currentIdx = 0;
        const states = [];
        const actions = [];

        function record(state, action) {
            // don't record if limit has been reached
            if (currentIdx >= limit) {
                return;
            }

            states[currentIdx] = state;
            actions[currentIdx] = action;
            currentIdx += 1;

            // clean forward history to forbid REDO upon changes
            states.splice(currentIdx);
            actions.splice(currentIdx);
        }

        function travel(state, distance = 1) {
            const targetIdx = currentIdx + distance;
            if (states[targetIdx] == null) {
                return state;
            }
            currentIdx = targetIdx;
            return states[currentIdx];
        }

        const reducer = (next, initialState) => (state, action) => {
            if (action.type === CommandActionType.UNDO_ACTION) {
                return travel(state, -(action.payload.distance));
            }
            if (action.type === CommandActionType.REDO_ACTION) {
                return travel(state, action.payload.distance);
            }
            if (recordedActions.includes(action.type)) {
                record(state, action);
            }
            return next(state, action);
        };

        return { reducer };
    }

I am then composing the `reducer` function together with application reducers.

Now if I dispatch some actions, it records all of them every time from beginning with new one added.  Also the `travel` function suffers even worst problem, all UNDO/REDO actions are replayed also causing to skip much further in history than wanted.

First issue can be avoided with check like `actions.includes(action)`, but it sounds bad enough to have a code just to cover devtools issue. However I don't see a way how to fix the second issue.

**Comment by gaearon** [comment]
I already explained how to fix this: please contribute tests so I feel safe making optimizations and can remove the re-evaluation. https://github.com/gaearon/redux-devtools/issues/81#issuecomment-133209702

**Comment by gaearon** [comment]
That said this reducer violates Redux constraints. It is not pure because it has local state (`states` and `actions`). That's why it doesn't work. If you move `state` and `actions` inside of its Redux state it will work as you'd like.

**Comment by FredyC** [comment]
Alright, fair point. It might be actually easier to have each state to refer to previous state it can be undo-ed to.

**Comment by gaearon** [comment]
Example unit test:

```js
import { createStore } from 'redux';
import devTools, { ActionCreators } from '../src/devTools';

function counter(state = 0, action) {
  switch (action.type) {
  case 'INCREMENT': return state + 1;
  case 'INCREMENT': return state - 1;
  default: return state;
  }
}

describe('devTools', () => {
  let store;
  let devToolsStore;

  beforeEach(() => {
    store = devTools()(createStore)(counter);
    devToolsStore = store.devToolsStore;
  });

  it('performs actions', () => {
    expect(store.getState()).toBe(0);
    store.dispatch({ type: 'INCREMENT' });
    expect(store.getState()).toBe(1);
    store.dispatch({ type: 'INCREMENT' });
    expect(store.getState()).toBe(2);
  });

  it('rollbacks state to the last committed state', () => {
    store.dispatch({ type: 'INCREMENT' });
    store.dispatch({ type: 'INCREMENT' });
    expect(store.getState()).toBe(2);

    devToolsStore.dispatch(ActionCreators.commit());
    expect(store.getState()).toBe(2);

    store.dispatch({ type: 'INCREMENT' });
    store.dispatch({ type: 'INCREMENT' });
    expect(store.getState()).toBe(4);

    devToolsStore.dispatch(ActionCreators.rollback());
    expect(store.getState()).toBe(2);

    store.dispatch({ type: 'DECREMENT' });
    expect(store.getState()).toBe(1);

    devToolsStore.dispatch(ActionCreators.rollback());
    expect(store.getState()).toBe(2);
  });
});
```

I hope this gives the idea!

**Comment by gaearon** [comment]
There's a PR attempted to fix it: https://github.com/gaearon/redux-devtools/pull/133
Can you please test whether it improves the situation?

**Comment by gaearon** [comment]
This is fixed in `redux-devtools@2.1.5`.
Performing an action with DevTools open was O(n), now it's O(1).
